### ISSUE TITLE
Support for wasm

### ISSUE THREAD

**Issue by trojek** [nats-io/nats.go#661]
## Feature Request
Support for WebAssembly

#### Use Case:
Use NATS from browser without using JS binding.

#### Proposed Change:
Add support for wasm.

#### Who Benefits From The Change(s)?
People who want to use NATS in WebAssembly. 

#### Alternative Approaches
Use JS binding.

**Comment by gedw99** [comment]
5the rust nats client or the golang client cant because they assume tcp ?
If those clients are modified to support web sockets it might work

**Comment by cameronbraid** [comment]
Another important use case is to write a Envoy Proxy Filter https://istio.io/latest/docs/concepts/wasm/

**Comment by john-livelyvideo** [comment]
any headway on this front?

**Comment by john-livelyvideo** [comment]
@derekcollison I'm simply trying to figure out if it's possible to run golang/webassembly in the web browser, while using the go client.

**Comment by derekcollison** [comment]
Not sure if that would be a good experience over the TS/JS client or not. I think you would need to bridge to JS for access to the websocket IIRC.

**Comment by john-livelyvideo** [comment]
All we need is a generic interface for sending/recv'ing websocket frames which works both natively and in the browser, right?

https://natsio.slack.com/archives/CM3T6T7JQ/p1652115060809699?thread_ts=1652109614.003509&cid=CM3T6T7JQ

**Comment by gedw99** [comment]
@derekcollison 

i also wanted to use the nats golang client from wasm to use in a browser. I really prefer to not have to mess with TS / JS client.

Can you explain why the experience would be bad ?  The golang client tries to do its own client side load balancing is its talking to a NATS Cluster , which is pretty cool. No need for cloud load balancers.

**Comment by derekcollison** [comment]
@gedw99 the TS/JS clients can do the same. I believe you would need to jump through to JS to access the websocket from the TS/JS client regardless. But I could be wrong.

**Comment by john-livelyvideo** [comment]
@derekcollison that is correct, but that's a pretty simple interop between golang/js, right?  per the framing i mentioned earlier?

**Comment by john-livelyvideo** [comment]
see the webassembly example in pion as a reference
https://github.com/pion/webrtc/blob/8c31eb98dd78c813d09deb249bdc977b83bff3f0/examples/README.md#webassembly

**Comment by derekcollison** [comment]
Will take a look, thanks.

**Comment by gedw99** [comment]
Hey all,

I am using the NATS Rountripper for third parties. Its great and works great with Connect. That's one use case sorted.
This however only support HTTP RPC. So its pull based

But i need my own GUI clients a way to talk to NATS to have 2 way communications.
I use GIOUI to build clients. Love it and make it easy and fast to build Golang GUI.
NATS.go works great from Desktop and Web over tcp.
NATS.go does NOT work form WASM Brower :) Hence why i am explaining this use case and ways we can build a test harness if we want.

Here is a nice and simple REST example written in Golang and using the GIOUI gui libs.
https://github.com/bpingris/gioui-requests. 
It will make a good Test harness for NATS Websockets and NATS TCP as well as NATS HttpRoundtriper.
You can compile this to Web ( WASM), Desktop, Mobile, Apple TV. 

If we compile to tinygo, I would be running WASM nats.go at the  Service Worker layer. 
Service workers have access to networking and indexeddb. 
They also run in the background, even when the Browser "tab" is closed, if the user installs as PWA. Recently IOS safari even supports it ( shock horror ).

here is the Chromium Test for it: https://chromium.googlesource.com/external/w3c/web-platform-tests/+/master/service-workers/service-worker/websocket-in-service-worker.https.html

**Comment by derekcollison** [comment]
I am supportive of getting the client to compile under TinyGO, or have another client targeting it. I believe the only thing is the encoded connections and GOB IIRC.

**Comment by gedw99** [comment]
To gui is rapidly catching up to the standard golang wasm compiler though.

For now we could get it working with the standard golang wasm compiler , which would mean it’s a larger file.

For wasm websockets there are many example of golang code that uses websockets and runs as wasm in the browser. 

Fetch with wasm to avoid net/http golang lib import shows this concept: https://github.com/marwan-at-work/wasm-fetch

the js calls are wrapped here: https://github.com/marwan-at-work/wasm-fetch/blob/master/fetch.go

the same thing should work for web socket / https3 WebTransport. Yes I know nats server does not yet support http3 WebTransport yet

**Comment by derekcollison** [comment]
There will be a TinyGo client in the future.

**Comment by gedw99** [comment]
https://github.com/oderwat/go-nats-app works now.

uses  https://github.com/nhooyr/websocket
